File: frontend/src/types/case.ts
export type ClassificationLabel = "legitimate" | "suspicious" | "impersonated" | "phishing" | "fraud";

export type ThreatCategory =
  | "legitimate"
  | "spam"
  | "phishing"
  | "spear_phishing"
  | "business_email_compromise"
  | "spoofing"
  | "credential_theft"
  | "malware_delivery"
  | "suspicious_unknown";

export type CaseStatus = "new" | "investigating" | "escalated" | "resolved" | "false_positive";

export type SignalSeverity = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";

export interface CaseSummary {
  case_id: number;
  email_analysis_id: number;
  subject: string;
  from_address: string;
  fraud_score: number;
  classification_label: ClassificationLabel;
  threat_category: ThreatCategory;
  status: CaseStatus;
  source_url: string;
  created_at: string;
}

export interface ReceivedHop {
  hop_index: number;
  raw: string;
  from_host: string | null;
  by_host: string | null;
  ip: string | null;
  timestamp: string | null;
}

export interface AuthCrossCheckEntry {
  upstream: string | null;
  ours: string;
  agreement: "match" | "mismatch" | "no_upstream_data";
}

export interface HeaderAuthResult {
  spf_result: string;
  spf_domain: string;
  dkim_result: string;
  dkim_domain: string;
  dmarc_result: string;
  dmarc_policy: string;
  spf_aligned: boolean;
  dkim_aligned: boolean;
  routing_anomaly_flags: string[];
  received_chain: ReceivedHop[];
  auth_cross_check: Record<string, AuthCrossCheckEntry>;
}

export interface DetectionResult {
  classification_label: ClassificationLabel;
  fraud_score: number;
  threat_category: ThreatCategory;
  secondary_indicators: string[];
}

export interface Signal {
  id: number;
  name: string;
  category: "authentication" | "identity" | "content" | "infrastructure" | "reputation" | "url";
  severity: SignalSeverity;
  score: number;
  explanation: string;
  evidence: string;
}

export interface ExtractedUrl {
  raw_url: string;
  anchor_text: string | null;
  normalized_url: string;
  hostname: string;
  registered_domain: string;
  scheme: string;
  is_ip_based: boolean;
  subdomain_count: number;
  has_punycode: boolean;
  is_shortener: boolean;
  display_text_mismatch: boolean;
  lookalike_of_brand: string | null;
  lexical_risk_score: number;
}

export interface DomainIntelligence {
  domain: string;
  registered_domain: string;
  mx_records: string[];
  nameservers: string[];
  registrar: string | null;
  created_date: string | null;
  age_days: number | null;
  dnsbl_listed: boolean;
  source: string;
  available: boolean;
  unavailable_reason: string | null;
}

export interface IpIntelligence {
  ip: string;
  ip_version: number;
  is_public: boolean;
  asn: string | null;
  asn_org: string | null;
  country: string | null;
  region: string | null;
  city: string | null;
  isp_org: string | null;
  is_vpn_or_proxy_or_tor: boolean | null;
  dnsbl_listed: boolean;
  source: string;
  available: boolean;
  unavailable_reason: string | null;
  disclaimer: string;
}

export interface ThreatIntelResult {
  provider: string;
  indicator_type: string;
  indicator_value: string;
  available: boolean;
  verdict: string | null;
}

export interface Attachment {
  filename: string;
  content_type: string;
  size_bytes: number;
  sha256: string;
}

export interface EmailAnalysisDetail {
  id: number;
  original_filename: string;
  source_url: string;
  subject: string;
  from_display_name: string;
  from_address: string;
  reply_to: string;
  return_path: string;
  to_addresses: string[];
  cc_addresses: string[];
  date_header: string;
  message_id: string;
  authentication_results: { raw: string[]; parsed_tokens: Record<string, string> };
  mime_structure: string[];
  body_text: string;
  attachments: Attachment[];
  raw_eml_sha256: string;
  created_at: string;
  header_auth_result: HeaderAuthResult | null;
  detection_result: DetectionResult | null;
  signals: Signal[];
  extracted_urls: ExtractedUrl[];
  domain_intelligence: DomainIntelligence[];
  ip_intelligence: IpIntelligence[];
  threat_intel_results: ThreatIntelResult[];
}

export interface CaseDetail {
  id: number;
  title: string;
  status: CaseStatus;
  overall_risk_level: ClassificationLabel | "unknown";
  created_at: string;
  email_analyses: EmailAnalysisDetail[];
}

export interface UploadResponse {
  case_id: number;
  email_analysis_id: number;
  status: string;
}

export interface TimelineEvent {
  id: number;
  case_id: number;
  email_analysis_id: number | null;
  event_type: string;
  description: string;
  occurred_at: string;
}

export interface CaseNote {
  id: number;
  case_id: number;
  author: string;
  note_text: string;
  created_at: string;
}

export interface EvidenceLogEntry {
  id: number;
  case_id: number;
  email_analysis_id: number | null;
  action: string;
  payload: Record<string, unknown>;
  prev_entry_hash: string;
  entry_hash: string;
  created_at: string;
}

export interface EvidenceLog {
  valid: boolean;
  first_broken_index: number | null;
  entries: EvidenceLogEntry[];
}

export interface GraphNode {
  id: string;
  type: "case" | "ip" | "domain";
  label: string;
}

export interface GraphEdge {
  source: string;
  target: string;
  relationship: "observed_in" | "potential_correlation";
}

export interface CaseGraph {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface SharedIndicator {
  indicator_type: string;
  indicator_value: string;
  other_case_ids: number[];
  note: string;
}

export interface DashboardStats {
  total_cases: number;
  high_risk_count: number;
  critical_count: number;
  potential_campaigns: number;
  investigations_today: number;
  classification_breakdown: Record<string, number>;
  threat_category_breakdown: Record<string, number>;
  recent_cases: {
    id: number;
    title: string;
    overall_risk_level: string;
    status: CaseStatus;
    created_at: string;
  }[];
}

File: backend/app/models/case.py
import enum
from datetime import datetime, timezone
from typing import Optional

from sqlalchemy import JSON, Column
from sqlmodel import Field, Relationship, SQLModel


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class ClassificationLabel(str, enum.Enum):
    """Matches the SIH26106 problem statement's own 5-way wording — this is the
    risk-band label graded against the PS, kept separate from ThreatCategory below."""

    legitimate = "legitimate"
    suspicious = "suspicious"
    impersonated = "impersonated"
    phishing = "phishing"
    fraud = "fraud"


class ThreatCategory(str, enum.Enum):
    """Investigator-facing threat taxonomy — a case can be risk-band 'fraud' *and*
    category 'business_email_compromise' at the same time; these answer different
    questions (how risky vs. what kind of attack)."""

    legitimate = "legitimate"
    spam = "spam"
    phishing = "phishing"
    spear_phishing = "spear_phishing"
    business_email_compromise = "business_email_compromise"
    spoofing = "spoofing"
    credential_theft = "credential_theft"
    malware_delivery = "malware_delivery"
    suspicious_unknown = "suspicious_unknown"


class CaseStatus(str, enum.Enum):
    new = "new"
    investigating = "investigating"
    escalated = "escalated"
    resolved = "resolved"
    false_positive = "false_positive"


class Case(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    title: str
    status: str = CaseStatus.new.value
    overall_risk_level: str = "unknown"
    created_at: datetime = Field(default_factory=_utcnow)

    email_analyses: list["EmailAnalysis"] = Relationship(back_populates="case")


class EmailAnalysis(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    case_id: int = Field(foreign_key="case.id")

    raw_eml_path: str
    raw_eml_sha256: str
    original_filename: str = ""
    # Deep link back to the original message (e.g. a Gmail URL) — set by the
    # Chrome extension on upload; empty for manually-uploaded .eml files.
    source_url: str = ""

    subject: str = ""
    from_display_name: str = ""
    from_address: str = ""
    reply_to: str = ""
    return_path: str = ""
    to_addresses: list = Field(default_factory=list, sa_column=Column(JSON))
    cc_addresses: list = Field(default_factory=list, sa_column=Column(JSON))
    date_header: str = ""
    message_id: str = ""

    raw_headers: dict = Field(default_factory=dict, sa_column=Column(JSON))
    authentication_results: dict = Field(default_factory=dict, sa_column=Column(JSON))
    mime_structure: list = Field(default_factory=list, sa_column=Column(JSON))
    body_text: str = ""
    body_html: str = ""
    attachments: list = Field(default_factory=list, sa_column=Column(JSON))

    created_at: datetime = Field(default_factory=_utcnow)

    case: Optional[Case] = Relationship(back_populates="email_analyses")
    header_auth_result: Optional["HeaderAuthResult"] = Relationship(
        back_populates="email_analysis",
        sa_relationship_kwargs={"uselist": False},
    )
    detection_result: Optional["DetectionResult"] = Relationship(
        back_populates="email_analysis",
        sa_relationship_kwargs={"uselist": False},
    )


class HeaderAuthResult(SQLModel, table=True):
    """Official PS component 2: header/protocol analysis."""

    id: Optional[int] = Field(default=None, primary_key=True)
    email_analysis_id: int = Field(foreign_key="emailanalysis.id", unique=True)

    spf_result: str = "none"
    spf_domain: str = ""
    dkim_result: str = "none"
    dkim_domain: str = ""
    dmarc_result: str = "none"
    dmarc_policy: str = ""
    spf_aligned: bool = False
    dkim_aligned: bool = False

    routing_anomaly_flags: list = Field(default_factory=list, sa_column=Column(JSON))
    received_chain: list = Field(default_factory=list, sa_column=Column(JSON))
    auth_cross_check: dict = Field(default_factory=dict, sa_column=Column(JSON))

    email_analysis: Optional[EmailAnalysis] = Relationship(back_populates="header_auth_result")


class DetectionResult(SQLModel, table=True):
    """Official PS component 1: fraudulent email detection engine. Per-signal
    evidence now lives in the Signal table (models/signal.py), not inline here."""

    id: Optional[int] = Field(default=None, primary_key=True)
    email_analysis_id: int = Field(foreign_key="emailanalysis.id", unique=True)

    classification_label: str = ClassificationLabel.legitimate.value
    fraud_score: int = 0
    threat_category: str = ThreatCategory.legitimate.value
    secondary_indicators: list = Field(default_factory=list, sa_column=Column(JSON))

    email_analysis: Optional[EmailAnalysis] = Relationship(back_populates="detection_result")
